### PR TITLE
Abhi/metrics manager

### PR DESCRIPTION
**Pull request by pretentious7**

### CODE REVIEW

File: plane/src/drone/backend_manager.rs [hunk 1]
```diff
@@ -84,12 +152,19 @@ impl BackendManager {
         state: BackendState,
         docker: PlaneDocker,
         state_callback: impl Fn(&BackendState) -> Result<(), Box<dyn Error>> + Send + Sync + 'static,
+        metrics_sender: TypedSocketSender<BackendMetricsMessage>,
```
Comment (paulgb) [inline plane/src/drone/backend_manager.rs:155]: the problem with this is that the `TypedSocketSender` is not durable across reconnects.

One alternative here is to pass the `Arc<StateStore>`, and then call `state_store.get_metrics_sender()` to get the sender for each message at send-time.

**Review by paulgb — COMMENTED** [review]
I like the approach. One correctness issue but I think with the way you've structured things it should be an easy fix (see comment).

File: plane/src/drone/docker/mod.rs [hunk 2]
```diff
@@ -166,6 +168,112 @@ impl PlaneDocker {
...
+
+    Ok(BackendMetricsMessage {
+        backend_id,
+        mem_total,
+        mem_active,
+        mem_inactive,
+        mem_unevictable,
+        mem_available,
+        cpu_used: container_cpu_used_delta,
+        sys_cpu: system_cpu_used_delta,
```
Comment (rolyatmax) [inline plane/src/drone/docker/mod.rs:273]: These messages previously had a `mem_used` value along with drone and cluster info. Is there a reason why we left those out? https://github.com/drifting-in-space/plane/blob/plane-0.3.x/core/src/messages/agent.rs#L168

Comment (pretentious7) [reply plane/src/drone/docker/mod.rs]: sorry, based this pr wrong, discussion here: https://github.com/drifting-in-space/plane/pull/543

File: plane/src/drone/backend_manager.rs (hunk 1 above)
Comment (pretentious7) [reply plane/src/drone/backend_manager.rs]: I've used the same metrics_sender as the one in state_store with an arc, does this work?

Comment (pretentious7) [reply plane/src/drone/backend_manager.rs]: oh wait a sec, it does not.

File: plane/src/drone/backend_manager.rs [hunk 3]
```diff
@@ -40,6 +42,73 @@ impl StepStatusResult {
...
+                    ) else {
+                        continue;
+                    };
+                    let Ok(_) = sender
+                        .read()
+                        .expect("failed to get read lock on metrics sender!")
+                        .send(metrics_message)
+                    else {
+                        continue;
+                    };
```
Comment (paulgb) [inline plane/src/drone/backend_manager.rs:103]: this has no effect, I think?

Comment (pretentious7) [reply plane/src/drone/backend_manager.rs]: ayup!

**Review by paulgb — APPROVED** [review]

**Pull request merged**